core/src/main/scala/kafka/api/FetchResponse.scala

```diff
@@ -48,6 +49,36 @@ object FetchResponsePartitionData {
 
 case class FetchResponsePartitionData(error: Short = Errors.NONE.code, hw: Long = -1L, messages: MessageSet) {
   val sizeInBytes = FetchResponsePartitionData.headerSize + messages.sizeInBytes
+
+  def toMessageFormat(toMagicValue: Byte): FetchResponsePartitionData = {
```

Comment:
I agree that it's better to do the message conversion in KafkaApis. The reason is that a single network thread is used to handle many socket connections. If the sending of one response is slow (e.g., recompression when converting the message), it slows down the processing of all connections on this network thread. On the other hand, if the processing in a request handler thread is slow, it only slows down that request. Other requests are unaffected.